### ISSUE TITLE
`layer:` missing features

### ISSUE THREAD

**Issue by ReinhardKeil** [Open-CMSIS-Pack/devtools#643]
There are a few missing features that could improve how `*.clayer.yml` files can be matched.  Below is a list (see also https://github.com/Open-CMSIS-Pack/devtools/blob/main/tools/projmgr/docs/Manual/YML-Input-Format.md#layer).

1. Introduce `for-board:` and `for-device:` as option to filter for a matching `device:` / `board:` specification.  This could be used instead of the meta-data in the *.pdsc files.

2. Allow to list `packs:` at the level of a `*.clayer.yml` file. With this feature it is just required to enter the `*.clayer.yml` file as a `variable:`.

3. The debug output of the csolution list layers command is quite useful, we might retain it (slightly reformatted). We might consider a  new option `-v` or  `-- verbose` that displays this kind of information.

4. During development of new `*.clayer.yml` files it might be useful to specify a search path for such layers.  This could be a command line option `y` or `--clayer-path` where all files of the type *.clayer.yml in this path are checked for matching meta-data (`layers:` - `type:` and/or `connections:`.

5. When layer `variables` in `*.csolution.yml` are not set, the `list layers` option should still work and identify the compatible layers.

----
**Workflow for users should be:**

`*.csolution.yml`/`*.cproject.yml` files provide a template project with layer types and layer variables.

1. User adds a new `target-type:` and selects a `board:` to `*.csolution.yml`
2. Execution of command `csolution list layers` outputs compatible layers for the empty layer variables.
3. User adds the identified layers as variables below the `target-type:` selection

```
  target-types:
    - type: frdmk22f_agmp03    # Added in step 1
      board: NXP::FRDM-K22F
      variables:                            # Added in step 2
        - Board-Layer:  ../../../../../boards/frdmk22f/frdmk22f.clayer.yml
        - Shield-Layer: ../../../../../shields/agmp03/agmp03.clayer.yml
    - type: lpcxpresso54114_agmp03
```
----
**Suggested Output without --verbose option**

```
csolution list layers -s fxas21002_fifo.csolution.yml
valid for context: fxas21002_fifo.*+frdmk22f_agmp03'    # output could be combined if identical for all build-types
 C:/w/NXP_Sensor_SDK/examples/issdk/sensors/fxas21002/fifo/fxas21002_fifo.cproject.yml
 C:/w/NXP_Sensor_SDK/boards/frdmk22f/frdmk22f.clayer.yml (layer type: Board)
 C:/w/NXP_Sensor_SDK/shields/agmp03/agmp03.clayer.yml (layer type: Shield)
    set: Bus.I2C (Jumper configuration: I2C/SPI=I2C, I2C=I2C0 - FXAS21002 I2C Bus)

valid for context: fxas21002_fifo.*+lpcxpresso54114_agmp03'
  C:/w/NXP_Sensor_SDK/examples/issdk/sensors/fxas21002/fifo/fxas21002_fifo.cproject.yml
  C:/w/NXP_Sensor_SDK/boards/lpcxpresso54114/lpcxpresso54114.clayer.yml (layer type: Board)
  C:/w/NXP_Sensor_SDK/shields/agmp03/agmp03.clayer.yml (layer type: Shield)
    set: Bus.I2C (Jumper configuration: I2C/SPI=I2C, I2C=I2C0 - FXAS21002 I2C Bus)
```

**Suggested Output with --verbose option**
Using this option, every possible combination (after `for-board:` and `for-device:` filtering) would be tested and shown.
```
check for context: fxas21002_fifo.debug+frdmk22f_agmp03' 
  C:/w/NXP_Sensor_SDK/examples/issdk/sensors/fxas21002/fifo/fxas21002_fifo.cproject.yml
  C:/w/NXP_Sensor_SDK/boards/frdmk22f/frdmk22f.clayer.yml (layer type: Board)
  C:/w/NXP_Sensor_SDK/shields/agmp03/agmp03.clayer.yml  (layer type: Shield)
    set: Bus.I2C (Jumper configuration: I2C/SPI=SPI, I2C=I2C0 - FXLS8962 I2C Bus)
required connections not provided:
  FXAS21002_I2C
  ARDUINO_UNO_I2C

check for context: fxas21002_fifo.debug+frdmk22f_agmp03' 
 C:/w/NXP_Sensor_SDK/examples/issdk/sensors/fxas21002/fifo/fxas21002_fifo.cproject.yml
 C:/w/NXP_Sensor_SDK/boards/frdmk22f/frdmk22f.clayer.yml (layer type: Board)
 C:/w/NXP_Sensor_SDK/shields/agmp03/agmp03.clayer.yml (layer type: Shield)
    set: Bus.I2C (Jumper configuration: I2C/SPI=I2C, I2C=I2C0 - FXAS21002 I2C Bus)
valid
```

----
When `*.clayer.yml` files are already specified, the tool should validate the combination and output the `set:` info to the user.
If there is no valid `set:` the user should be notified, example:

**Suggested Output without --verbose option**

```
csolution list layers -s fxas21002_fifo.csolution.yml
valid for context: fxas21002_fifo.*+frdmk22f_agmp03'    # output could be combined if identical for all build-types
 C:/w/NXP_Sensor_SDK/examples/issdk/sensors/fxas21002/fifo/fxas21002_fifo.cproject.yml
 C:/w/NXP_Sensor_SDK/boards/frdmk22f/frdmk22f.clayer.yml (layer type: Board)
 C:/w/NXP_Sensor_SDK/shields/agmp03/agmp03.clayer.yml (layer type: Shield)
    set: Bus.I2C (Jumper configuration: I2C/SPI=I2C, I2C=I2C0 - FXAS21002 I2C Bus)

invalid for context: fxas21002_fifo.*+lpcxpresso54114_agmp03'
  C:/w/NXP_Sensor_SDK/examples/issdk/sensors/fxas21002/fifo/fxas21002_fifo.cproject.yml
  C:/w/NXP_Sensor_SDK/boards/lpcxpresso54114/lpcxpresso54114.clayer.yml (layer type: Board)
  C:/w/NXP_Sensor_SDK/shields/agmp03/agmp03.clayer.yml (layer type: Shield)
```

**Comment by RobertRostohar** [comment]
A few additional items:

1. `connections:` sharing same `set:` should also be combined. In the following example each sensor should be avaialbe separately but also combined for either I2C or SPI.

```
  connections:
    - connect: Sensor1 I2C Bus
      set: Bus.I2C
      info: "Jumper configuration: I2C/SPI=I2C"
      provides:
        - SENSOR1_I2C
      consumes:
        - ARDUINO_UNO_I2C
    - connect: Sensor1 SPI Bus
      set: Bus.SPI
      info: "Jumper configuration: I2C/SPI=SPI"
      provides:
        - SENSOR1_SPI
      consumes:
        - ARDUINO_UNO_SPI
    - connect: Sensor2 I2C Bus
      set: Bus.I2C
      info: "Jumper configuration: I2C/SPI=I2C"
      provides:
        - SENSOR2_I2C
      consumes:
        - ARDUINO_UNO_I2C
    - connect: Sensor2 SPI Bus
      set: Bus.SPI
      info: "Jumper configuration: I2C/SPI=SPI"
      provides:
        - SENSOR2_SPI
      consumes:
        - ARDUINO_UNO_SPI

```

  The following combinations should be checked:
   - Bus.I2C: provides: SENSOR1_I2C
   - Bus.I2C: provides: SENSOR2_I2C
   - Bus.I2C: provides: SENSOR1_I2C and SENSOR2_I2C (currently not considered)
   - Bus.SPI: provides: SENSOR1_SPI
   - Bus.SPI: provides: SENSOR2_SPI
   - Bus.SPI: provides: SENSOR1_SPI and SENSOR2_SPI (currently not considered)

2. Searching for `layers:` specified with `type:` is limited to `packs:` specified in the solution. Consider the example below. It would make sense to search through all installed packs to find compatible layers.

```
  layers:
    - type: Board
    - type: Shield
```

**Comment by brondani** [comment]
@RobertRostohar

1. I am not sure if I understand the motivation, but I am afraid that implicitly combining nodes leads to worse readability of the connections.
What is the purpose of creating several `connect` nodes with the same `set` values in a single layer?
Isn't the code below easier to read while still providing the proper validation?
``` yml
  connections:
    - connect: Sensors I2C Bus
      set: Bus.I2C
      info: "Jumper configuration: I2C/SPI=I2C"
      provides:
        - SENSOR1_I2C
        - SENSOR2_I2C
      consumes:
        - ARDUINO_UNO_I2C

    - connect: Sensors SPI Bus
      set: Bus.SPI
      info: "Jumper configuration: I2C/SPI=SPI"
      provides:
        - SENSOR1_SPI
        - SENSOR2_SPI
      consumes:
        - ARDUINO_UNO_SPI
```

2. Concerning the `packs` being filtered when they are explicitly listed, this is the default behaviour for all commands so I would keep it for consistency.
For searching through all installed packs the `--load=all` option can be used:
```
-l, --load arg         Set policy for packs loading [latest|all|required]
```

**Comment by RobertRostohar** [comment]
1. `connections:` sharing same `set:`
> What is the purpose of creating several connect nodes with the same set values in a single layer?

To have more granularity in the requirements and enable combinations which would not be possible otherwise.

My original example was not complete. I have left out SPI CS lines for simplicity. Consider the following example:

```
  connections:
     - connect: Sensor1 SPI Bus
      set: Bus.SPI
      info: "Jumper configuration: I2C/SPI=SPI"
      provides:
        - SENSOR1_SPI
        - SENSOR1_CS
      consumes:
        - ARDUINO_UNO_SPI
        - ARDUINO_UNO_D2
    - connect: Sensor2 SPI Bus
      set: Bus.SPI
      info: "Jumper configuration: I2C/SPI=SPI"
      provides:
        - SENSOR2_SPI
        - SENSOR2_CS
      consumes:
        - ARDUINO_UNO_SPI
        - ARDUINO_UNO_D3
```
Combining `SENSOR_1` and `SENSOR_2 `which use the same `set: Bus.SPI` would imply that besides `SPI` also `D2` and `D3` are required. If not both `D2` and `D3` are provided, none of the `SENSOR_1` and `SENSOR_2` can be used. But the following combinations should still be valid:
- `SENSOR_1`: `SPI` + `D2`
- `SENSOR_2`: `SPI` + `D3`

2. Using `--load=all` should be fine. I was not aware of that option. Thanks!

**Comment by iomint** [comment]
one comment related to " 1 -Introduce for-board: and for-device: as option to filter for a matching device: / board: specification. This could be used instead of the meta-data in the *.pdsc files."
I would prefer to not duplicate the filtering capabilities in two different ways in the clayer or in the *.pdsc files

In a broader way, concerning all the proposals aiming to connect Sw to Hw  like 'target-type', 'connect', 'set', 'provide', .... I am more in favor of a Hw project concept dual of the software project relying on the hw parts description introduced in [173](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/issues/173) and [174](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/issues/174) and the possibility to express conditions to link Sw components with Hw parts as proposed in  [182](https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/issues/182)

**Comment by ReinhardKeil** [comment]
@iomint two different boards could use the same set of <parts>.  How would you express that one layer is configured for board1 and another for board2?

**Comment by brondani** [comment]
The requested features were implemented in the above mentioned PRs and the following additional refinements were done:
- The `-c, --context` command line option accepts wildcards `*, ?` so now it's possible to select multiple contexts when running a command.
- For consistency the `list layers` command prints a plain list of included layers (referenced by valid path) and compatible layers (referenced by type) into the stdout for all selected contexts.
With the `-v,--verbose` option an additional Info message with a summary of valid configurations (project, layers and possible exclusive `set` combinations) is printed into stdout for each selected context.
With the `-d,--debug` option a Debug message with the full validation of connections combinations (valid and invalid) is printed into the stderr for each selected context.

**Comment by ReinhardKeil** [comment]
I did some tests with https://github.com/brondani/cmsis-toolbox/releases/tag/1.5.0-dev1 and it looks good. Suggest to add a command "get clayer" with almost identical syntax, but output that can be directly copied into the `*.csolution.yml`.

For example:
```
csolution get layers -s .\fxas21002_fifo.csolution.yml -L c:\w\NXP_Sensor_SDK

valid configuration #1 for target type:  frdmk22f_agmp03 
      variables:
          - Board-Layer:  C:/w/NXP_Sensor_SDK/boards/boards/frdmk22f/frdmk22f.clayer.yml
          - Shield-Layer: C:/w/NXP_Sensor_SDK/shields/agmp03/agmp03.clayer.yml
#          set: Bus.I2C-Alt (Jumper configuration: I2C/SPI=I2C, I2C=I2C1 - FXLS8962 Alternate I2C Bus)
#          set: Bus.I2C-Alt (Jumper configuration: I2C/SPI=I2C, I2C=I2C1 - FXAS21002 Alternate I2C Bus)
#          set: Bus.I2C-Alt (Jumper configuration: I2C/SPI=I2C, I2C=I2C1 - MPL3115 Alternate I2C Bus)
#          set: Bus.I2C-Alt (Jumper configuration: I2C/SPI=I2C, I2C=I2C1 - MAG3110 Alternate I2C Bus)

valid configuration #2 for target type:  frdmk22f_agmp03 
 ...
```

Another minor improvement: Add on warning output the context information:
```
$Board-Layer$ - warning csolution: variable not defined for context 'fxas21002_fifo.debug+frdmk22f_agmp03'
$Shield-Layer$ - warning csolution: variable not defined for context 'fxas21002_fifo.debug+frdmk22f_agmp03'
$Board-Layer$ - warning csolution: variable not defined for context 'fxas21002_fifo.release+frdmk22f_agmp03'
$Shield-Layer$ - warning csolution: variable not defined for context 'fxas21002_fifo.release+frdmk22f_agmp03'
```

**Comment by jkrech** [comment]
This is now officially released in CMSIS-Toolbox 1.5.0
